imap_processing/spice/geometry.py

```diff
@@ -149,13 +156,58 @@ def get_spacecraft_spin_phase(
     spin_phase : float or np.ndarray
         Spin phase for the input query times.
     """
-    if isinstance(query_met_times, float):
-        # TODO: call get_spin_data function to get spin data for the
-        # input query times
-        # Here, return a float (dummy implementation)
-        return 0.0  # Replace this with actual logic to calculate spin phase
-
-    # Return an ndarray of the same shape, filled with 0.0 for
-    # now (dummy implementation)
-    # TODO: Replace with actual logic to calculate spin phase
-    return np.array(query_met_times, dtype=float)
+    spin_df = get_spin_data()
+
+    # Ensure query_met_times is an array
+    query_met_times = np.asarray(query_met_times)
+    is_scalar = query_met_times.ndim == 0
+    if is_scalar:
+        # Force scalar to array because np.asarray() will not
+        # convert scalar to array
+        query_met_times = np.atleast_1d(query_met_times)
+
+    # Create an empty array to store spin phase results
+    spin_phases = np.zeros_like(query_met_times)
+
+    # Find all spin time that are less or equal to query_met_times.
+    # To do that, use side right, a[i-1] <= v < a[i], in the searchsorted.
+    # Eg.
+    # >>> df['a']
+    # array([0, 15, 30, 45, 60])
+    # >>> np.searchsorted(df['a'], [0, 13, 15, 32, 70], side='right')
+    # array([1, 1, 2, 3, 5])
+    last_spin_indices = np.searchsorted(
+        spin_df["spin_start_time"], query_met_times, side="right"
+    )
+    # Use np.clip to ensure that the last_spin_indices are within the range.
+    # >>> np.clip(x, 0, len(df['a']) -1 )
+    # array([1, 1, 2, 3, 4])
+    last_spin_indices = np.clip(last_spin_indices, 0, len(spin_df) - 1)
```

Comment:
I thought we were going to raise if this happened?

```suggestion
    # Make sure we aren't going out of the bounds
    if query_met_times.min() < spin_df["spin_start_time"][0] or query_met_times.max() > spin_df["spin_start_time"][-1] + spin_df["spin_period_sec"][-1]:
        raise ValueError
```

Reply:
That works. I hope we add checks in dependency to avoid this value error.